flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/metrics/MetricFetcher.java

```diff
@@ -0,0 +1,185 @@
+								toRetain.add(job.getJobId().toString());
+							}
+							for (JobDetails job : details.getFinishedJobs()) {
+								toRetain.add(job.getJobId().toString());
+							}
+							synchronized (metrics) {
+								metrics.jobs.keySet().retainAll(toRetain);
+							}
+						}
+					}, ctx);
+
+				/**
+				 * Requests the metric dump from the job manager.
+				 */
+				jobManager.ask(Messages.getRequestMetrics(), timeout)
+					.onSuccess(new OnSuccess<Object>() {
+						@Override
+						public void onSuccess(Object result) throws Throwable {
+							addMetrics(result);
+						}
+					}, ctx);
+
+				/**
+				 * We first request the list of all registered task managers from the job manager, and then
+				 * request the respective metric dump from each task manager.
+				 *
+				 * All stored metrics that do not belong to a registered task manager will be removed.
+				 */
+				jobManager.ask(JobManagerMessages.getRequestRegisteredTaskManagers(), timeout)
+					.onSuccess(new OnSuccess<Object>() {
+						@Override
+						public void onSuccess(Object result) throws Throwable {
+							Iterable<Instance> taskManagers = ((JobManagerMessages.RegisteredTaskManagers) result).asJavaIterable();
+							List<String> activeTaskManagers = new ArrayList<>();
+							for (Instance taskManager : taskManagers) {
+								activeTaskManagers.add(taskManager.getId().toString());
+								taskManager.getActorGateway().ask(Messages.getRequestMetrics(), timeout)
+									.onSuccess(new OnSuccess<Object>() {
+										@Override
+										public void onSuccess(Object result) throws Throwable {
+											addMetrics(result);
+										}
+									}, ctx);
+							}
+							synchronized (metrics) { // remove all metrics belonging to unregistered task managers
+								metrics.taskManagers.keySet().retainAll(activeTaskManagers);
+							}
+						}
+					}, ctx);
+			}
+		} catch (Exception e) {
+			LOG.warn("Exception while fetching metrics.", e);
+		}
+	}
+
+	private void addMetrics(Object result) {
+		Object[] data = (Object[]) result;
+		for (int x = 0; x < data.length; ) {
+			synchronized (metrics) {
+				switch ((byte) data[x++]) {
+					case 0:
+					case 1:
+						String name = (String) data[x++];
+						Object value = data[x++];
+						metrics.add(name, value);
+						break;
+					case 2:
+						String histogramName = (String) data[x++];
+						metrics.add(histogramName + "_min", data[x++]);
+						metrics.add(histogramName + "_max", data[x++]);
+						metrics.add(histogramName + "_mean", data[x++]);
+						metrics.add(histogramName + "_median", data[x++]);
+						metrics.add(histogramName + "_stddev", data[x++]);
+						metrics.add(histogramName + "_p75", data[x++]);
+						metrics.add(histogramName + "_p90", data[x++]);
+						metrics.add(histogramName + "_p95", data[x++]);
+						metrics.add(histogramName + "_p98", data[x++]);
+						metrics.add(histogramName + "_p99", data[x++]);
+						metrics.add(histogramName + "_p999", data[x++]);
```

Comment:
This is probably quite error prone, because whenever the histogram is changed (e.g. order of serialized fields) it has to be reflected here as well. I think it would be better if we would have a serializer which does the serialization and deserialization in one place.